pennylane/templates/subroutines/qdrift.py

```diff
@@ -0,0 +1,257 @@
+            qml.Hadamard(0)
+
+            # Evolve according to H
+            qml.QDrift(H, time=1.2, n = 10)
+
+            # Measure some quantity
+            return qml.probs()
+
+    >>> my_circ()
+    [0.71061676 0.         0.28938324 0.        ]
+
+
+    .. details::
+        :title: Usage Details
+
+        We currently **Do NOT** support computing gradients with respect to the
+        coefficients of the input Hamiltonian. We can however compute the gradient
+        with respect to the evolution time:
+
+        .. code-block:: python3
+
+            dev = qml.device("default.qubit", wires=2)
+
+            @qml.qnode(dev)
+            def my_circ(time):
+                # Prepare H:
+                H = qml.dot([0.2, -0.1], [qml.PauliY(0), qml.PauliZ(1)])
+
+                # Prepare some state
+                qml.Hadamard(0)
+
+                # Evolve according to H
+                qml.QDrift(H, time, n=10, seed=10)
+
+                # Measure some quantity
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+
+        >>> args = qnp.array([1.23])
+        >>> print(qml.grad(my_circ)(*tuple(args)))
+        0.27980654844422853
+    """
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self, hamiltonian, time, n=1, seed=None, decomposition=None, id=None
+    ):
+        r"""Initialize the QDrift class"""
+
+        if isinstance(hamiltonian, Hamiltonian):
+            coeffs, ops = hamiltonian.terms()
+
+        elif isinstance(hamiltonian, Sum):
+            coeffs, ops = [], []
+            for op in hamiltonian:
+                try:
+                    coeffs.append(op.scalar)
+                    ops.append(op.base)
+                except AttributeError:  # coefficient is 1.0
+                    coeffs.append(1.0)
+                    ops.append(op)
+
+        else:
+            raise TypeError(
+                f"The given operator must be a PennyLane ~.Hamiltonian or ~.Sum got {hamiltonian}"
+            )
+
+        if len(ops) < 2:
+            raise ValueError(
+                "There should be atleast 2 terms in the Hamiltonian. Otherwise use `qml.exp`"
+            )
+
+        if any(requires_grad(coeff) for coeff in coeffs):
+            raise qml.QuantumFunctionError(
+                "The QDrift template currently doesn't support differentiation through the "
+                "coefficients of the input Hamiltonian."
+            )
+
+        if decomposition is None:  # need to do this to allow flatten and _unflatten
+            unwrapped_coeffs = unwrap(coeffs)
+            decomposition = _sample_decomposition(unwrapped_coeffs, ops, time, n=n, seed=seed)
+
+        self._hyperparameters = {
+            "n": n,
+            "seed": seed,
+            "base": hamiltonian,
+            "decomposition": decomposition,
+        }
+        super().__init__(time, wires=hamiltonian.wires, id=id)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
```

Comment:
Do we actually need this method?

Reply:
Yes, because the hamiltonian is not a trainable parameter, so we need to pass it in carefully.

Reply:
Sure but since you are including the hamiltonian as kwarg in the init, by default it is assumed untrainable (I think)

Reply:
But the init method expects it as a arg instead of a kwarg.